Source/player.cpp

```diff
@@ -514,26 +514,44 @@ void StartSpell(int pnum, Direction d, int cx, int cy)
 		return;
 	}
 
+	// Checks conditions for spell again, cause initial check was done when spell was queued and the parameters could be changed meanwhile
+	bool isValid = true;
+	switch (player.queuedSpell.spellType) {
+	case RSPLTYPE_SKILL:
+	case RSPLTYPE_SPELL:
+		isValid = CheckSpell(MyPlayerId, player.queuedSpell.spellId, player.queuedSpell.spellType, true) == SpellCheckResult::Success;
+		break;
+	case RSPLTYPE_SCROLL:
+		isValid = UseScroll(player, player.queuedSpell.spellId);
+		break;
+	case RSPLTYPE_CHARGES:
+		isValid = UseStaff(player, player.queuedSpell.spellId);

```

Comment:
Does `CMD_CHANGEPLRITEMS` also set the current charges of the staff correctly?
If not this could also result in a desync, cause Client A check's correctly that there are not enough charges and forbids the cast. Client B sees more charges and still casts.

Reply:
It doesn't. Since the staff is regenerated from the item seed, it will actually just give it the charges it would have at spawn. Also, the `UseStaff()` function doesn't reduce staff charges. That happens in `UseMana()` when it calls `UseStaffCharge()`, but `UseMana()` doesn't do anything if the casting player isn't the local player so staff charges aren't tracked accurately at all.

But I don't consider that to be much of an issue. Assuming Client A is the caster in your scenario, they won't send the network message in the first place since they don't have charges left to cast. That type of desync could only happen as a result of queuing the spell on Client A and then cancelling it as a result of this logic, which will only happen once at the point when the staff is depleted.